Toolchain/roboctl-0.3.8/Commands/Legoctl/legoctl.1
.TH LEGOCTL 1
.SH NAME    \" Section header
.PP
 
legoctl - Lego robot communication and control utility

\" Underline anything that is typed verbatim - commands, etc.
.SH SYNOPSIS
.PP
.nf 
.na 
legoctl [flags] command [arguments]
    
Enter "legoctl" with no arguments for a brief list of flags.
.ad
.fi

\" Optional sections
.SH "PURPOSE"

The
.B legoctl
command is a utility for communicating with and controlling Lego robotics
bricks, such as the RCX, and NXT.

.SH "DESCRIPTION"

.B Legoctl
is used to perform basic operations such as uploading files, displaying
status information about the brick, starting and stopping programs, etc.

.B Legoctl
is based on the Robot Control library (-lroboctl), and is intended to
run on all Unix-like compatible platforms.  It was developed under FreeBSD,
and is also known to work under Cygwin, Linux, and Mac OS X.

Since the default security settings of typical Unix systems are fairly
restrictive, it may be necessary to adjust device permissions to allow
non root users to use
.B legoctl.

.SH "USB on FreeBSD"

On FreeBSD 5.x and later, this requires altering the
.B devfs
configuration.

For Lego NXT USB devices, the simplest solution (for those who aren't
concerned about the security of their USB devices) is adding the following
lines to /etc/devfs.rules:

.nf
.na
    [system=10]
    add path 'ugen*' mode 666 group operator
    add path 'usb*' mode 666 group operator
.ad
.fi

Also make sure /etc/rc.conf contains the following:

.nf
.na
    devd_enable="YES"
    devfs_system_ruleset="system"
.ad
.fi

Then run:

.nf
.na
    /etc/rc.d/devfs restart
.ad
.fi

USB device nodes are created according to these rules when the device
is plugged in.
Note that these rules make all generic USB devices writable to all users.

A slightly more secure method for systems with multiple users would be
to use permissions of 664, and add the those users whom you wish to
have write access to the devices to the operator group.

.SH "USB on Linux (Xubuntu 6)"

Note that these are the configuration steps necessary for Xubuntu 6.
Other Linux distributions may require additional configuration.

Like FreeBSD, Linux USB device nodes are created on-the-fly when a device is
connected, and are by default only writable by root.
To make the NXT writable by all users, create the file
/etc/udev/rules.d/70-legonxt.rules containing the following lines:

.nf
.na
# Lego NXT
BUS="usb", SYSFS{idVendor}="0694", MODE="0666" 
.ad
.fi 

.SH "Serial ports on FreeBSD"

Lego RCX users will need to configure a serial port for brick
communication.  Serial device nodes are static (created at boot time)
on FreeBSD.  To make them writable use chmod
to change the permissions on /dev/cuaa*, and edit /etc/devfs.conf to
make the changes survive the next reboot.
For RCX serial devices, the following rules would do the trick:

.nf
.na
    perm    /dev/cuaa0  0666
    perm    /dev/cuaa1  0666
.ad
.fi

Note again that these rules make all serial devices plugged into com1 and
com2 writable to all users.  Consider the consequences before setting up
such rules.

.SH "BLUETOOTH"

The Bluetooth interface is currently supported on FreeBSD and Linux.
Before you can use bluetooth, you must configure the bluetooth daemon
for your OS, and then pair the NXT with your computer.  ("Pairing" a
bluetooth device is like plugging in a wired device
such as USB, Firewire, etc.)

To configure and pair your bluetooth connection in a fairly automated
fashion, run the 
.B lego_bluetooth_config
script.  ( Included with the roboctl package. )
If you prefer to do the configuration
manually, or if lego_bluetooth_config
does not yet support your platform, see the instructions below.

.SH "BLUETOOTH ON FREEBSD"

To learn more about bluetooth options on FreeBSD, please see the Bluetooth
section in the FreeBSD handbook.

Note: These are the configuration steps followed by the lego_bluetooth_config
script.  You can perform these steps manually, or simply run the script.

First, the appropriate bluetooth kernel modules must be loaded.  For
most USB bluetooth dongles, this is done using

.nf
.na
    kldload ng_ubt
.ad
.fi

To load the module at boot time, add the following to /boot/loader.conf:

.nf
.na
    ng_ubt_load="YES"
.ad
.fi

You can scan for bluetooth devices using

.nf
.na
    hccontrol inquiry
.ad
.fi

If you have more than one bluetooth channel, you can specify one for
any hccontrol command with the -n flag, e.g.

.nf
.na
    hccontrol -n ubt0hci inquiry
.ad
.fi

The names of reporting devices can be retrieved using the
remote_name_request command with the bluetooth address, e.g.:

.nf
.na
    hccontrol remote_name_request 00:11:22:33:44:55 0 0 0
.ad
.fi

(Substitute your device address for 00:11:22:33:44:55).

Next, you will need to enable the
.B hcsecd
daemon.
This requires two steps.  First, add an entry like the following
to /etc/bluetooth/hcsecd.conf:

.nf
.na
    device {
	    bdaddr  00:11:22:33:44:55;
	    name    "NXT";
	    key     nokey;
	    pin     "1234";
    }
.ad
.fi

Note that "NXT" is the default bluetooth name of the NXT brick, and
1234 is the default passcode.  Naturally, if you have changed these
settings on the Lego controller, you must configure hcsecd.conf to
match.

Next, add the following to /etc/rc.conf:

.nf
.na
    hcsecd_enable="YES"
.ad
.fi

To activate hcsecd, you will need to reboot or run 

.nf
.na
    /etc/rc.d/hcsecd start
.ad
.fi

With hcsecd running, you can now perform pairing of your NXT with
your FreeBSD box using the interface on the NXT brick.  Note that
the brick may report "Line busy" when you connect.  This is a normal
glitch, but will not prevent the link for working properly.  If your
Lego controller finds your computer and asks you for a passcode when
you connect, you have probably succeeded.

Finally, add an entry like the following to /etc/bluetooth/hosts:

00:11:22:33:44:55   NXT

By default,
.B Legoctl
(and all other programs based on the roboctl library)
will try to look up the address of the bluetooth device named "NXT"
in the system bluetooth hosts database using 
.B bt_gethostbyname(3).
Note that the name in the bluetooth hosts database does not have to be
the same name programmed into the NXT brick.  This is simply a name
the operating system associates with the bluetooth address.

If you have more than one NXT, they will of course need different
names in the hosts database.
You can select a non-default name on the command line with

--btname name

or by setting the environment variable
.B ROBOCTL_BTNAME.

In any case, the name and corresponding address must be in the bluetooth
hosts database on your system.  See
.B lego_bluetooth_config
above.

.SH "BLUETOOTH ON LINUX"

The methods outlined here have been tested on Xubuntu 6.06.  Xubuntu
has the bluetooth interface preconfigured and running in the default
installation.  Note that this may not be the case with all Linux 
distributions.

These instructions currently work only for root.  Additional instructions
for making bluetooth devices available to non-privileged users will be
added when time permits.

You can scan for bluetooth devices using

.nf
.na
    hcitool scan
.ad
.fi

You can perform pairing of your NXT with your Linux box using the interface
on the NXT brick (recommended), or manipulate the pairing from Linux
using the following commands:

.nf
.na
hcitool scan                    # List devices in range
hcitool cc 00:11:22:33:44:55    # Create connection
.ad
.fi

.SH FILES
.nf
.na
/dev/usb*, /dev/ugen*, /etc/devfs.conf, /etc/devfs.rules, /etc/usbd.conf
/etc/bluetooth/hcsecd.conf, ~/.legoctl/bluetooth_address,
/boot.loader.conf, /etc/rc.conf
.ad
.fi

.SH ENVIRONMENT
There are currently no environment settings that directly affect
.B legoctl.

.SH "SEE ALSO"
nbc(1), nxc(1), nqc(1), roboctl(3), vexctl(1), ape(1), devfs(8), hcsecd(8)

.SH BUGS
Many commands are not yet implemented.

.SH EXAMPLES
.nf
.na
legoctl status
legoctl --btname NXT2 status
legoctl upload prog.rxe
.ad
.fi

.SH AUTHOR
.nf
.na
Jason W. Bacon
http://acadix.biz



Toolchain/roboctl-0.3.8/Commands/Legoctl/legoctl.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sysexits.h>
#include <roboctl.h>
#include "legoctl_cmd.h"
#include "protos.h"

extern int     Debug;

int     main(int argc,char *argv[])

{
    arg_t   arg_data = {"",NULL,0,0};
    rct_cmd_t   cmd = RCT_CMD_STATUS;
    unsigned int    flags = RCT_PROBE_DEV_ALL;

    switch(argc)
    {
	case    1:
	    legoctl_usage(argv[0]);
	    break;
	default:
	    break;
    }

    if ( parse_args(argc,argv,&cmd,&arg_data,&flags) == RCT_OK )
    {
	return legoctl(cmd,&arg_data,flags);
    }
    else
	return EX_USAGE;
}


int     legoctl(rct_cmd_t cmd,arg_t *arg_data,unsigned int flags)

{
    rct_brick_list_t    bricks;

    rct_find_bricks(&bricks,arg_data->bluetooth_name,RCT_PROBE_ALL);
    debug_printf("Found %d bricks...\n",rct_brick_count(&bricks));
    if ( rct_brick_count(&bricks) == 0 )
    {
	fputs("Sorry, no accessible bricks found.\n",stderr);
	exit(EX_UNAVAILABLE);
    }
    
    switch(cmd)
    {
	case    RCT_CMD_STATUS:
	case    RCT_CMD_STOP:
	    return multi_brick_cmd(&bricks,cmd,flags);
	case    RCT_CMD_UPLOAD:
	case    RCT_CMD_DELETE:
	case    RCT_CMD_START:
	case    RCT_CMD_PLAY_SOUND:
	    return file_cmd(&bricks,arg_data->filename,cmd,flags);
	case    RCT_CMD_PLAY_TONE:
	    return play_tone(&bricks,arg_data->herz,arg_data->milliseconds);
	case    RCT_CMD_DOWNLOAD:
	case    RCT_CMD_FIRM_UP:
	case    RCT_CMD_FIRM_DOWN:
	    fputs("This command is not yet implemented.\n",stderr);
	    break;
	default:
	    fprintf(stderr,"Internal error: Invalid command: %d.\n",cmd);
	    fputs("This is a program bug.  Please report it to the author(s).\n",stderr);
	    exit(EX_SOFTWARE);
    }
    return 0;
}


int     multi_brick_cmd(rct_brick_list_t *bricks,rct_cmd_t cmd,unsigned int flags)

{
    rct_brick_t         *brick;
    int     c;
    
    for (c=0; c<rct_brick_count(bricks); ++c)
    {
	brick = rct_get_brick_from_list(bricks,c);
	if ( rct_open_brick(brick) == RCT_OK )
	{
	    switch(cmd)
	    {
		case    RCT_CMD_STATUS:
		    /* Print device info */
		    if ( rct_print_device_info(brick) != 0 )
			return 1;
		    
		    /* Print firmware version */
		    rct_print_firmware_version(brick);
		    
		    /* Print battery level */
		    rct_print_battery_level(brick);
		    break;
		case    RCT_CMD_STOP:
		    rct_stop_program(brick);
		    break;
		default:
		    break;
	    }
	    
	    rct_close_brick(rct_get_brick_from_list(bricks,c));
	}
	else
	{
	    fprintf(stderr,"Error opening brick.\n");
	    return 1;
	}
    }
    return 0;
}


int     file_cmd(rct_brick_list_t *bricks,char *filename,rct_cmd_t cmd,
		    unsigned int flags)

{
    rct_brick_t *brick;
    rct_flag_t  loop;
    
    switch(rct_brick_count(bricks))
    {
	case    0:
	    fputs("Sorry, accessible no bricks found.\n",stderr);
	    exit(EX_UNAVAILABLE);
	case    1:
	    brick = rct_get_brick_from_list(bricks,0);
	    if ( rct_open_brick(brick) == RCT_OK )
	    {
		switch(cmd)
		{
		    case    RCT_CMD_START:
			rct_start_program(brick,filename);
			break;
		    case    RCT_CMD_UPLOAD:
			if ( flags & RCT_OVERWRITE )
			    rct_delete_file(brick,filename);
			rct_upload_file(brick,filename,RCT_UPLOAD_PLAY_SOUND);
			break;
		    case    RCT_CMD_DELETE:
			rct_delete_file(brick,filename);
			break;
		    case    RCT_CMD_PLAY_SOUND:
			loop = ((flags & RCT_LOOP) != 0);
			rct_play_sound_file(brick,loop,filename);
			break;
		    default:
			break;
		}
		rct_close_brick(brick);
	    }
	    break;
	default:
	    fputs("Error: multiple bricks connected.  Don't know which one to upload to.\n",stderr);
	    exit(EX_DATAERR);
    }
    return 0;
}


int     play_tone(rct_brick_list_t *bricks,int herz,int milliseconds)

{
    rct_brick_t *brick;
    
    switch(rct_brick_count(bricks))
    {
	case    0:
	    fputs("Sorry, accessible no bricks found.\n",stderr);
	    exit(EX_UNAVAILABLE);
	case    1:
	    brick = rct_get_brick_from_list(bricks,0);
	    if ( rct_open_brick(brick) == RCT_OK )
	    {
		rct_play_tone(brick,herz,milliseconds);
		rct_close_brick(brick);
	    }
	    break;
	default:
	    fputs("Error: multiple bricks connected.  Don't know which one to upload to.\n",stderr);
	    exit(EX_DATAERR);
    }
    return 0;
}


void    legoctl_usage(char *progname)

{
    fputs("Usage:\n",stderr);
    fprintf(stderr,"\t%s [flags] status\n",progname);
    fprintf(stderr,"\t%s [flags] upload <filename> [slot #]\n",progname);
    fprintf(stderr,"\t%s [flags] playsound <filename>\n",progname);
    fprintf(stderr,"\t%s [flags] playtone <herz> <milliseconds>\n",progname);
    //fprintf(stderr,"\t%s [flags] download <filename|all> [slot #]\n",progname);
    fprintf(stderr,"\t%s [flags] delete <filename> [slot #]\n",progname);
    //fprintf(stderr,"\t%s [flags] firmware_up <filename>\n",progname);
    //fprintf(stderr,"\t%s [flags] firmware_down <filename>\n",progname);
    fprintf(stderr,"\t%s [flags] start <filename|slot #>\n",progname);
    fprintf(stderr,"\t%s [flags] stop\n",progname);
    fputs("\nFlags:\n",stderr);
    //fputs("\t--nxt       probe for NXT only\n",stderr);
    //fputs("\t--usb       probe USB busses only\n",stderr);
    //fputs("\t--bluetooth probe bluetooth interfaces only\n",stderr);
    //fputs("\t--rcx       probe for RCX only\n",stderr);
    fputs("\t--overwrite overwrite existing files on brick\n",stderr);
    fputs("\t--loop      repeat command indefinitely\n",stderr);
    fputs("\t--debug     enable debugging output\n",stderr);
    fputs("\t--btname name  Specify a non-default bluetooth name\n", stderr);
    exit(EX_USAGE);
}


int     parse_args(int argc,char *argv[],rct_cmd_t *cmd,arg_t *arg_data,
	unsigned int *flags)

{
    int     arg;
    char    *end;
    
    for (arg = 1; argv[arg] != NULL; ++arg)
    {
	if ( strcmp(argv[arg],"status") == 0 )
	{
	    *cmd = RCT_CMD_STATUS;
	}
	else if ( strcmp(argv[arg],"upload") == 0 )
	{
	    *cmd = RCT_CMD_UPLOAD;
	    /* The next argument should be the last */
	    if ( arg == argc - 2 )
		arg_data->filename = argv[++arg];
	    else
		legoctl_usage(argv[0]);
	}
	else if ( strcmp(argv[arg],"playsound") == 0 )
	{
	    *cmd = RCT_CMD_PLAY_SOUND;
	    /* The next argument should be the last */
	    if ( arg == argc - 2 )
		arg_data->filename = argv[++arg];
	    else
		legoctl_usage(argv[0]);
	}
	else if ( strcmp(argv[arg],"playtone") == 0 )
	{
	    *cmd = RCT_CMD_PLAY_TONE;
	    /* The next argument should be the last */
	    if ( arg == argc - 3 )
	    {
		arg_data->herz = strtol(argv[++arg],&end,0);
		if ( end == argv[arg] )
		{
		    fprintf(stderr,"Invalid herz value: '%s'\n",argv[arg]);
		    legoctl_usage(argv[0]);
		}
		arg_data->milliseconds = strtol(argv[++arg],&end,0);
		if ( end == argv[arg] )
		{
		    fprintf(stderr,"Invalid milliseconds value: '%s'\n",argv[arg]);
		    legoctl_usage(argv[0]);
		}
	    }
	    else
		legoctl_usage(argv[0]);
	}
	else if ( strcmp(argv[arg],"download") == 0 )
	{
	    *cmd = RCT_CMD_DOWNLOAD;
	}
	else if ( strcmp(argv[arg],"firmware_up") == 0 )
	{
	    *cmd = RCT_CMD_FIRM_UP;
	    /* The next argument should be the last */
	    /*if ( arg == argc - 2 )
		arg_data->filename = argv[++arg];
	    else
		legoctl_usage(argv[0]);
	    */
	}
	else if ( strcmp(argv[arg],"firmware_down") == 0 )
	{
	    *cmd = RCT_CMD_FIRM_DOWN;
	}
	else if ( strcmp(argv[arg],"delete") == 0 )
	{
	    *cmd = RCT_CMD_DELETE;
	    /* The next argument should be the last */
	    if ( arg == argc - 2 )
		arg_data->filename = argv[++arg];
	    else
		legoctl_usage(argv[0]);
	}
	else if ( strcmp(argv[arg],"start") == 0 )
	{
	    *cmd = RCT_CMD_START;
	    /* The next argument should be the last */
	    if ( arg == argc - 2 )
		arg_data->filename = argv[++arg];
	    else
		legoctl_usage(argv[0]);
	}
	else if ( strcmp(argv[arg],"stop") == 0 )
	{
	    *cmd = RCT_CMD_STOP;
	}
	else if ( strcmp(argv[arg],"--nxt") == 0 )
	{
	    *flags |= RCT_PROBE_DEV_NXT;
	}
	else if ( strcmp(argv[arg],"--rcx") == 0 )
	{
	    *flags |= RCT_PROBE_DEV_RCX;
	}
	else if ( strcmp(argv[arg],"--overwrite") == 0 )
	{
	    *flags |= RCT_OVERWRITE;
	}
	else if ( strcmp(argv[arg],"--loop") == 0 )
	{
	    *flags |= RCT_LOOP;
	}
	else if ( strcmp(argv[arg],"--debug") == 0 )
	{
	    Debug = 1;
	}
	else if ( strcmp(argv[arg],"--btname") == 0 )
	{
	    arg_data->bluetooth_name = argv[++arg];
	}
	else
	{
	    fprintf(stderr,"Invalid option: %s\n",argv[arg]);
	    legoctl_usage(argv[0]);
	}
    }
    return RCT_OK;
}

